awsenv/replacer.go

```diff
@@ -1,89 +1,138 @@
 package awsenv
 
 import (
+	"context"
 	"os"
-	"strings"
 
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/service/ssm"
 	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
 )
 
-// ParameterGetter is implemented by the ssm client
-type ParameterGetter interface {
-	GetParameter(*ssm.GetParameterInput) (*ssm.GetParameterOutput, error)
+var (
+	setenv  = os.Setenv
+	environ = os.Environ
+)
+
+// DefaultPrefix holds the standard environment value prefix.
+var DefaultPrefix = "awsenv:"
+
+// ParamsGetter represents a data source that can translate parameter names
+// or paths into parameter values.
+type ParamsGetter interface {
+	GetParams(ctx context.Context, names []string) (map[string]string, error)
 }
 
-// Replacer handles replacing existing environment variables with parameter values
-type Replacer struct {
-	ssm    ParameterGetter
-	prefix string
+// LimitedParamsGetter represents a ParamsGetter that can describe its own
+// request limit. A ParamsGetter implementing this interface will not be
+// given more names per request than the number returned by GetParamsLimit,
+// unless it returns a value <= 0, which will be interpreted as unlimited.
+type LimitedParamsGetter interface {
+	ParamsGetter
+	GetParamsLimit() int
 }
 
 // NewReplacer returns a Replacer that will operate on env vars with the
-// given value prefix, using the given ParameterGetter.
-func NewReplacer(envValuePrefix string, ssm ParameterGetter) *Replacer {
+// given value prefix, using the given ParamsGetter.
+//
+// NewReplacer will panic if envValuePrefix is the empty string.
+func NewReplacer(envValuePrefix string, ssm ParamsGetter) *Replacer {
+	if envValuePrefix == "" {
+		panic("awsenv: envValuePrefix must be non-empty")
```

Comment:
Why not fallback to `DefaultPrefix`?

Reply:
the question would be: if `envValuePrefix == ""`, is it likely intentional, or due to some initialization error? e.g. if they meant for it to be `"blah!"` but it ended up as "", should we interpret it as "awsenv:" or fail loudly?

Reply:
Hmm, I suppose it would be better to fail loudly in this case. I'm okay with leaving it as is.